File: mobile/android/app/src/main/java/com/railgun/android/nodemode/routing/BloomFilter.kt
package com.railgun.android.nodemode.routing

import java.nio.ByteBuffer
import java.nio.ByteOrder
import java.security.MessageDigest
import java.util.BitSet
import java.util.UUID
import kotlin.math.ceil
import kotlin.math.ln
import kotlin.math.pow
import kotlin.math.roundToInt

/**
 * Space-efficient probabilistic data structure for bundle deduplication.
 * 
 * Used in epidemic routing to prevent message loops - peers exchange bloom
 * filters to efficiently determine which bundles the other already has.
 */
class BloomFilter(
    expectedElements: Int = 10_000,
    falsePositiveRate: Double = 0.01
) {
    
    // MARK: - Properties
    
    private val bits: BitSet
    val bitCount: Int
    val hashCount: Int
    var itemCount: Int = 0
        private set
    val createdAt: Long = System.currentTimeMillis()
    
    init {
        require(expectedElements > 0) { "Expected elements must be positive" }
        require(falsePositiveRate in 0.0..1.0) { "False positive rate must be between 0 and 1" }
        
        // Calculate optimal bit count: m = -n * ln(p) / (ln(2))^2
        val n = expectedElements.toDouble()
        val p = falsePositiveRate
        val m = -n * ln(p) / ln(2.0).pow(2)
        
        // Calculate optimal hash count: k = (m/n) * ln(2)
        val k = (m / n) * ln(2.0)
        
        bitCount = ceil(m).toInt()
        hashCount = maxOf(1, k.roundToInt())
        bits = BitSet(bitCount)
    }
    
    /**
     * Create a bloom filter from serialized data
     */
    constructor(data: ByteArray) : this(10_000, 0.01) {
        require(data.size >= 32) { "Invalid bloom filter data" }
        
        val buffer = ByteBuffer.wrap(data).order(ByteOrder.BIG_ENDIAN)
        
        val storedBitCount = buffer.getInt()
        val storedHashCount = buffer.getInt()
        val storedItemCount = buffer.getInt()
        // Skip createdAt (8 bytes) and padding (4 bytes)
        buffer.position(buffer.position() + 12)
        
        // Verify compatibility
        require(storedBitCount == bitCount && storedHashCount == hashCount) {
            "Incompatible bloom filter parameters"
        }
        
        itemCount = storedItemCount
        
        // Read bit array
        val wordCount = (bitCount + 63) / 64
        for (i in 0 until wordCount) {
            if (buffer.remaining() < 8) break
            val word = buffer.getLong()
            for (bit in 0 until 64) {
                if ((word and (1L shl bit)) != 0L) {
                    val bitIndex = i * 64 + bit
                    if (bitIndex < bitCount) {
                        bits.set(bitIndex)
                    }
                }
            }
        }
    }
    
    // MARK: - Computed Properties
    
    /**
     * Current estimated false positive rate
     */
    val estimatedFalsePositiveRate: Double
        get() {
            if (itemCount == 0) return 0.0
            val k = hashCount.toDouble()
            val n = itemCount.toDouble()
            val m = bitCount.toDouble()
            return (1.0 - Math.exp(-k * n / m)).pow(k)
        }
    
    /**
     * Fill ratio (proportion of bits set to 1)
     */
    val fillRatio: Double
        get() = bits.cardinality().toDouble() / bitCount
    
    // MARK: - Operations
    
    /**
     * Add a string to the bloom filter
     */
    fun add(item: String) {
        add(item.toByteArray(Charsets.UTF_8))
    }
    
    /**
     * Add bytes to the bloom filter
     */
    fun add(item: ByteArray) {
        val hashes = computeHashes(item)
        for (hash in hashes) {
            val index = (hash % bitCount.toLong()).toInt().let { if (it < 0) it + bitCount else it }
            bits.set(index)
        }
        itemCount++
    }
    
    /**
     * Add a UUID to the bloom filter
     */
    fun add(uuid: UUID) {
        add(uuid.toString())
    }
    
    /**
     * Check if an item might be in the set
     * Returns false if definitely not in set, true if possibly in set
     */
    fun mightContain(item: String): Boolean {
        return mightContain(item.toByteArray(Charsets.UTF_8))
    }
    
    /**
     * Check if bytes might be in the set
     */
    fun mightContain(item: ByteArray): Boolean {
        val hashes = computeHashes(item)
        for (hash in hashes) {
            val index = (hash % bitCount.toLong()).toInt().let { if (it < 0) it + bitCount else it }
            if (!bits.get(index)) {
                return false
            }
        }
        return true
    }
    
    /**
     * Check if a UUID might be in the set
     */
    fun mightContain(uuid: UUID): Boolean {
        return mightContain(uuid.toString())
    }
    
    /**
     * Union two bloom filters (combine their sets)
     */
    fun union(other: BloomFilter) {
        require(bitCount == other.bitCount && hashCount == other.hashCount) {
            "Bloom filters must have compatible parameters"
        }
        bits.or(other.bits)
        itemCount = maxOf(itemCount, other.itemCount)
    }
    
    /**
     * Clear all items
     */
    fun clear() {
        bits.clear()
        itemCount = 0
    }
    
    // MARK: - Serialization
    
    /**
     * Serialize bloom filter to bytes
     */
    fun serialize(): ByteArray {
        val wordCount = (bitCount + 63) / 64
        val buffer = ByteBuffer.allocate(32 + wordCount * 8).order(ByteOrder.BIG_ENDIAN)
        
        // Header
        buffer.putInt(bitCount)
        buffer.putInt(hashCount)
        buffer.putInt(itemCount)
        buffer.putLong(createdAt)
        buffer.putInt(0) // padding
        
        // Bit array as 64-bit words
        for (i in 0 until wordCount) {
            var word = 0L
            for (bit in 0 until 64) {
                val bitIndex = i * 64 + bit
                if (bitIndex < bitCount && bits.get(bitIndex)) {
                    word = word or (1L shl bit)
                }
            }
            buffer.putLong(word)
        }
        
        return buffer.array()
    }
    
    // MARK: - Hash Functions
    
    /**
     * Compute multiple hash values using double hashing
     * h_i(x) = h1(x) + i * h2(x)
     */
    private fun computeHashes(data: ByteArray): LongArray {
        // Use SHA-256 to get two 64-bit hash values
        val digest = MessageDigest.getInstance("SHA-256")
        val hash = digest.digest(data)
        
        // Split into two 64-bit values
        val hash1 = ByteBuffer.wrap(hash, 0, 8).order(ByteOrder.BIG_ENDIAN).getLong()
        val hash2 = ByteBuffer.wrap(hash, 8, 8).order(ByteOrder.BIG_ENDIAN).getLong()
        
        // Generate k hashes using double hashing
        return LongArray(hashCount) { i ->
            hash1 + i * hash2
        }
    }
    
    companion object {
        /**
         * Calculate optimal parameters for a bloom filter
         */
        fun optimalParameters(expectedElements: Int, falsePositiveRate: Double): Pair<Int, Int> {
            val n = expectedElements.toDouble()
            val p = falsePositiveRate
            val m = -n * ln(p) / ln(2.0).pow(2)
            val k = (m / n) * ln(2.0)
            return Pair(ceil(m).toInt(), maxOf(1, k.roundToInt()))
        }
    }
}

/**
 * A bloom filter that automatically scales by adding new filters
 * when the false positive rate exceeds a threshold
 */
class ScalableBloomFilter(
    private val initialCapacity: Int = 10_000,
    private val targetFalsePositiveRate: Double = 0.01,
    private val growthFactor: Double = 2.0
) {
    private val filters = mutableListOf<BloomFilter>()
    
    init {
        filters.add(BloomFilter(initialCapacity, targetFalsePositiveRate))
    }
    
    val itemCount: Int
        get() = filters.sumOf { it.itemCount }
    
    val filterCount: Int
        get() = filters.size
    
    fun add(item: String) {
        add(item.toByteArray(Charsets.UTF_8))
    }
    
    fun add(item: ByteArray) {
        // Check if current filter is getting full
        val lastFilter = filters.last()
        if (lastFilter.estimatedFalsePositiveRate > targetFalsePositiveRate) {
            // Add a new filter with larger capacity
            val newCapacity = (initialCapacity * growthFactor.pow(filters.size.toDouble())).toInt()
            val newFPRate = maxOf(targetFalsePositiveRate * 0.5.pow(filters.size.toDouble()), 0.001)
            filters.add(BloomFilter(newCapacity, newFPRate))
        }
        
        filters.last().add(item)
    }
    
    fun add(uuid: UUID) {
        add(uuid.toString())
    }
    
    fun mightContain(item: String): Boolean {
        return mightContain(item.toByteArray(Charsets.UTF_8))
    }
    
    fun mightContain(item: ByteArray): Boolean {
        return filters.any { it.mightContain(item) }
    }
    
    fun mightContain(uuid: UUID): Boolean {
        return mightContain(uuid.toString())
    }
    
    fun clear() {
        filters.clear()
        filters.add(BloomFilter(initialCapacity, targetFalsePositiveRate))
    }
}

File: mobile/ios/NodeMode/Sources/NodeMode/Routing/BloomFilter.swift
//
//  BloomFilter.swift
//  RailGun Node Mode
//
//  Space-efficient probabilistic data structure for bundle deduplication
//

import Foundation
import CommonCrypto

// MARK: - Bloom Filter

/// A space-efficient probabilistic set membership data structure.
/// Used for deduplication in epidemic routing - prevents message loops.
public struct BloomFilter {
    
    // MARK: - Properties
    
    /// The bit array
    private var bits: [UInt64]
    
    /// Number of bits in the filter
    public let bitCount: Int
    
    /// Number of hash functions
    public let hashCount: Int
    
    /// Number of items added
    public private(set) var itemCount: Int = 0
    
    /// Timestamp when created
    public let createdAt: Int64
    
    // MARK: - Computed Properties
    
    /// Current estimated false positive rate
    public var estimatedFalsePositiveRate: Double {
        let k = Double(hashCount)
        let n = Double(itemCount)
        let m = Double(bitCount)
        
        guard n > 0 else { return 0 }
        
        return pow(1.0 - exp(-k * n / m), k)
    }
    
    /// Fill ratio (proportion of bits set to 1)
    public var fillRatio: Double {
        var setCount = 0
        for word in bits {
            setCount += word.nonzeroBitCount
        }
        return Double(setCount) / Double(bitCount)
    }
    
    // MARK: - Initialization
    
    /// Create a bloom filter with specified parameters
    /// - Parameters:
    ///   - expectedElements: Expected number of elements to store
    ///   - falsePositiveRate: Desired false positive rate (0.01 = 1%)
    public init(expectedElements: Int = 10_000, falsePositiveRate: Double = 0.01) {
        precondition(expectedElements > 0, "Expected elements must be positive")
        precondition(falsePositiveRate > 0 && falsePositiveRate < 1, "False positive rate must be between 0 and 1")
        
        // Calculate optimal bit count: m = -n * ln(p) / (ln(2))^2
        let n = Double(expectedElements)
        let p = falsePositiveRate
        let m = -n * log(p) / pow(log(2), 2)
        
        // Calculate optimal hash count: k = (m/n) * ln(2)
        let k = (m / n) * log(2)
        
        self.bitCount = Int(m.rounded(.up))
        self.hashCount = max(1, Int(k.rounded()))
        
        // Initialize bit array (using 64-bit words)
        let wordCount = (bitCount + 63) / 64
        self.bits = [UInt64](repeating: 0, count: wordCount)
        
        self.createdAt = Int64(Date().timeIntervalSince1970 * 1000)
    }
    
    /// Create from serialized data
    public init(data: Data) throws {
        guard data.count >= 24 else {
            throw BloomFilterError.invalidData
        }
        
        var offset = 0
        
        // Read header
        let bitCount = data.subdata(in: offset..<(offset + 8))
            .withUnsafeBytes { $0.load(as: Int.self) }
        offset += 8
        
        let hashCount = data.subdata(in: offset..<(offset + 8))
            .withUnsafeBytes { $0.load(as: Int.self) }
        offset += 8
        
        let itemCount = data.subdata(in: offset..<(offset + 8))
            .withUnsafeBytes { $0.load(as: Int.self) }
        offset += 8
        
        let createdAt = data.subdata(in: offset..<(offset + 8))
            .withUnsafeBytes { $0.load(as: Int64.self) }
        offset += 8
        
        // Calculate expected word count
        let wordCount = (bitCount + 63) / 64
        let expectedDataSize = 32 + (wordCount * 8)
        
        guard data.count >= expectedDataSize else {
            throw BloomFilterError.invalidData
        }
        
        // Read bit array
        var bits = [UInt64]()
        bits.reserveCapacity(wordCount)
        
        for _ in 0..<wordCount {
            let word = data.subdata(in: offset..<(offset + 8))
                .withUnsafeBytes { $0.load(as: UInt64.self) }
            bits.append(word)
            offset += 8
        }
        
        self.bitCount = bitCount
        self.hashCount = hashCount
        self.itemCount = itemCount
        self.createdAt = createdAt
        self.bits = bits
    }
    
    // MARK: - Operations
    
    /// Add an item to the bloom filter
    public mutating func add(_ item: String) {
        add(item.data(using: .utf8)!)
    }
    
    /// Add an item to the bloom filter
    public mutating func add(_ item: Data) {
        let hashes = computeHashes(item)
        
        for hash in hashes {
            let index = hash % UInt64(bitCount)
            let wordIndex = Int(index / 64)
            let bitIndex = Int(index % 64)
            bits[wordIndex] |= (1 << bitIndex)
        }
        
        itemCount += 1
    }
    
    /// Add a UUID to the bloom filter
    public mutating func add(_ uuid: UUID) {
        add(uuid.uuidString)
    }
    
    /// Check if an item might be in the set
    /// Returns false if definitely not in set, true if possibly in set
    public func mightContain(_ item: String) -> Bool {
        mightContain(item.data(using: .utf8)!)
    }
    
    /// Check if an item might be in the set
    public func mightContain(_ item: Data) -> Bool {
        let hashes = computeHashes(item)
        
        for hash in hashes {
            let index = hash % UInt64(bitCount)
            let wordIndex = Int(index / 64)
            let bitIndex = Int(index % 64)
            
            if (bits[wordIndex] & (1 << bitIndex)) == 0 {
                return false
            }
        }
        
        return true
    }
    
    /// Check if a UUID might be in the set
    public func mightContain(_ uuid: UUID) -> Bool {
        mightContain(uuid.uuidString)
    }
    
    /// Union two bloom filters (combine their sets)
    public mutating func union(with other: BloomFilter) throws {
        guard bitCount == other.bitCount && hashCount == other.hashCount else {
            throw BloomFilterError.incompatibleFilters
        }
        
        for i in 0..<bits.count {
            bits[i] |= other.bits[i]
        }
        
        // Item count is approximate after union
        itemCount = max(itemCount, other.itemCount)
    }
    
    /// Clear all items
    public mutating func clear() {
        for i in 0..<bits.count {
            bits[i] = 0
        }
        itemCount = 0
    }
    
    // MARK: - Serialization
    
    /// Serialize bloom filter to Data
    public func serialize() -> Data {
        var data = Data()
        
        // Header
        withUnsafeBytes(of: bitCount) { data.append(contentsOf: $0) }
        withUnsafeBytes(of: hashCount) { data.append(contentsOf: $0) }
        withUnsafeBytes(of: itemCount) { data.append(contentsOf: $0) }
        withUnsafeBytes(of: createdAt) { data.append(contentsOf: $0) }
        
        // Bit array
        for word in bits {
            withUnsafeBytes(of: word) { data.append(contentsOf: $0) }
        }
        
        return data
    }
    
    // MARK: - Hash Functions
    
    /// Compute multiple hash values using double hashing
    /// h_i(x) = h1(x) + i * h2(x)
    private func computeHashes(_ data: Data) -> [UInt64] {
        // Use SHA-256 to get two 128-bit hash values
        var hash = [UInt8](repeating: 0, count: 32)
        data.withUnsafeBytes { ptr in
            _ = CC_SHA256(ptr.baseAddress, CC_LONG(data.count), &hash)
        }
        
        // Split into two 64-bit values
        let hash1 = hash[0..<8].withUnsafeBytes { $0.load(as: UInt64.self) }
        let hash2 = hash[8..<16].withUnsafeBytes { $0.load(as: UInt64.self) }
        
        // Generate k hashes using double hashing
        var hashes = [UInt64]()
        hashes.reserveCapacity(hashCount)
        
        for i in 0..<hashCount {
            let combined = hash1 &+ UInt64(i) &* hash2
            hashes.append(combined)
        }
        
        return hashes
    }
}

// MARK: - Bloom Filter Error

public enum BloomFilterError: Error, LocalizedError {
    case invalidData
    case incompatibleFilters
    
    public var errorDescription: String? {
        switch self {
        case .invalidData:
            return "Invalid bloom filter data"
        case .incompatibleFilters:
            return "Bloom filters have incompatible parameters"
        }
    }
}

// MARK: - Scalable Bloom Filter

/// A bloom filter that automatically scales by adding new filters
/// when the false positive rate exceeds a threshold
public class ScalableBloomFilter {
    
    // MARK: - Properties
    
    private var filters: [BloomFilter] = []
    private let initialCapacity: Int
    private let targetFalsePositiveRate: Double
    private let growthFactor: Double
    
    /// Total number of items across all filters
    public var itemCount: Int {
        filters.reduce(0) { $0 + $1.itemCount }
    }
    
    /// Number of internal filters
    public var filterCount: Int {
        filters.count
    }
    
    // MARK: - Initialization
    
    public init(
        initialCapacity: Int = 10_000,
        falsePositiveRate: Double = 0.01,
        growthFactor: Double = 2.0
    ) {
        self.initialCapacity = initialCapacity
        self.targetFalsePositiveRate = falsePositiveRate
        self.growthFactor = growthFactor
        
        // Create initial filter
        filters.append(BloomFilter(
            expectedElements: initialCapacity,
            falsePositiveRate: falsePositiveRate
        ))
    }
    
    // MARK: - Operations
    
    /// Add an item to the bloom filter
    public func add(_ item: String) {
        add(item.data(using: .utf8)!)
    }
    
    /// Add an item to the bloom filter
    public func add(_ item: Data) {
        // Check if current filter is getting full
        if let lastFilter = filters.last,
           lastFilter.estimatedFalsePositiveRate > targetFalsePositiveRate {
            // Add a new filter with larger capacity
            let newCapacity = Int(Double(initialCapacity) * pow(growthFactor, Double(filters.count)))
            let newFPRate = targetFalsePositiveRate * pow(0.5, Double(filters.count))
            filters.append(BloomFilter(
                expectedElements: newCapacity,
                falsePositiveRate: max(newFPRate, 0.001)
            ))
        }
        
        // Add to the last filter
        filters[filters.count - 1].add(item)
    }
    
    /// Add a UUID to the bloom filter
    public func add(_ uuid: UUID) {
        add(uuid.uuidString)
    }
    
    /// Check if an item might be in the set
    public func mightContain(_ item: String) -> Bool {
        mightContain(item.data(using: .utf8)!)
    }
    
    /// Check if an item might be in the set
    public func mightContain(_ item: Data) -> Bool {
        // Check all filters (any positive is a positive)
        for filter in filters {
            if filter.mightContain(item) {
                return true
            }
        }
        return false
    }
    
    /// Check if a UUID might be in the set
    public func mightContain(_ uuid: UUID) -> Bool {
        mightContain(uuid.uuidString)
    }
    
    /// Clear all filters
    public func clear() {
        filters.removeAll()
        filters.append(BloomFilter(
            expectedElements: initialCapacity,
            falsePositiveRate: targetFalsePositiveRate
        ))
    }
}
